test/python/circuit/library/test_linear_function.py

```diff
@@ -0,0 +1,224 @@
+from qiskit.circuit.library.generalized_gates import LinearFunction
+from qiskit.circuit.exceptions import CircuitError
+
+from qiskit.quantum_info.operators import Operator
+
+
+def random_linear_circuit(num_qubits, num_gates, seed=None):
+    """Generate a pseudo random linear circuit."""
+
+    gates = ["cx", "swap"]
+
+    instructions = {
+        "cx": (CXGate(), 2),
+        "swap": (SwapGate(), 2),
+    }
+
+    if isinstance(seed, np.random.Generator):
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+
+    samples = rng.choice(gates, num_gates)
+
+    circ = QuantumCircuit(num_qubits)
+
+    for name in samples:
+        gate, nqargs = instructions[name]
+        qargs = rng.choice(range(num_qubits), nqargs, replace=False).tolist()
+        circ.append(gate, qargs)
+
+    return circ
+
+
+def random_invertible_binary_matrix(num_qubits, seed=None):
+    """Generates a random invertible n x n binary matrix."""
+
+    # This code is adapted from random_cnotdihedral
+    if isinstance(seed, np.random.Generator):
+        rng = seed
+    else:
+        rng = np.random.default_rng(seed)
+
+    det = 0
+    while np.allclose(det, 0) or np.allclose(det, 2):
+        binary_matrix = rng.integers(2, size=(num_qubits, num_qubits))
+        det = np.linalg.det(binary_matrix) % 2
+
+    return binary_matrix
+
+
+@ddt
+class TestLinearFunctions(QiskitTestCase):
+    """Tests for clifford append gate functions."""
+
+    @data(2, 3, 4, 5, 6, 7, 8)
+    def test_conversion_to_matrix_and_back(self, num_qubits):
+        """Test correctness of first constructing a linear function from a linear quantum circuit,
+        and then synthesizing this linear function to a quantum circuit."""
+        rng = np.random.default_rng(1234)
+
+        for _ in range(10):
+            for num_gates in [0, 5, 5 * num_qubits]:
+                # create a random linear circuit
+                linear_circuit = random_linear_circuit(num_qubits, num_gates, seed=rng)
+                self.assertTrue(isinstance(linear_circuit, QuantumCircuit))
+
+                # convert it to a linear function
+                linear_function = LinearFunction(linear_circuit, validate_input=True)
+
+                # check that the internal matrix has right dimensions
+                self.assertTrue(linear_function.linear.shape == (num_qubits, num_qubits))
+
+                # synthesize linear function
+                synthesized_linear_function = linear_function.definition
+                self.assertTrue(isinstance(synthesized_linear_function, QuantumCircuit))
+
+                # check that the synthesized linear function only contains CX and SWAP gates
+                for inst, _, _ in synthesized_linear_function.data:
+                    self.assertTrue(isinstance(inst, (CXGate, SwapGate)))
+
+                # check equivalence to the original function
+                self.assertEqual(Operator(linear_circuit), Operator(synthesized_linear_function))
```

Comment:
It might be good to not only check equivalence, but also that the synthesis has the same or fewer instructions in it - I think it should be perfectly efficient, right, so that's a requirement as well.